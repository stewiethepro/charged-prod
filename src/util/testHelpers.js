import React from 'react';
import mapValues from 'lodash/mapValues';

import * as reactTestingLibrary from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { HelmetProvider } from 'react-helmet-async';

import configureStore from '../store';
import { IntlProvider } from './reactIntl';
import defaultConfig from '../config/configDefault';
import { ConfigurationProvider } from '../context/configurationContext';
import { RouteConfigurationProvider } from '../context/routeConfigurationContext';
import routeConfiguration from '../routing/routeConfiguration';

// In case you have translated the template and have new translations that
// are missing from the en translations file, the language for the tests can
// be changed here so that there are no missing translation keys in tests.
import messages from '../translations/en.json';

let undefined;
export const getDefaultConfiguration = () => {
  // TODO: add more relevant data for tests
  // TODO: make it possible to overwrite configuration for tests
  return {
    ...defaultConfig,
    currency: 'USD',
    marketplaceName: 'MarketplaceX',
    facebookAppId: undefined,
    marketplaceRootURL: 'http://localhost:3000',
    maps: {
      ...defaultConfig.maps,
      mapboxAccessToken: undefined,
      googleMapsAPIKey: undefined,
    },
    stripe: {
      ...defaultConfig.stripe,
      publishableKey: 'pk_test_',
    },
    listing: {
      listingFields: [
        {
          key: 'category',
          scope: 'public',
          includeForListingTypes: [
            'product-selling',
            'daily-booking',
            'nightly-booking',
            'hourly-booking',
          ],
          schemaType: 'enum',
          enumOptions: [
            { option: 'city-bikes', label: 'City bikes' },
            { option: 'e-bikes', label: 'E-bikes' },
            { option: 'mtb', label: 'MTB' },
            { option: 'kids-bikes', label: 'Kids bikes' },
          ],
          filterConfig: {
            indexForSearch: true,
            filterType: 'SelectSingleFilter',
            label: 'Category',
            group: 'primary',
          },
          showConfig: {
            label: 'Category',
            isDetail: true,
          },
          saveConfig: {
            label: 'Select category',
            placeholderMessage: 'Choose…',
            isRequired: true,
            requiredMessage: 'You need to select a category.',
          },
        },

        {
          key: 'tire-size',
          scope: 'public',
          includeForListingTypes: [
            'product-selling',
            'daily-booking',
            'nightly-booking',
            'hourly-booking',
          ],
          schemaType: 'enum',
          enumOptions: [
            { option: 29, label: '29' },
            { option: 28, label: '28' },
            { option: 27, label: '27' },
            { option: 26, label: '26' },
            { option: 24, label: '24' },
            { option: 20, label: '20' },
            { option: 18, label: '18' },
          ],
          filterConfig: {
            indexForSearch: true,
            label: 'Tire size',
            group: 'secondary',
          },
          showConfig: {
            label: 'Tire size',
            isDetail: true,
          },
          saveConfig: {
            label: 'Tire size',
            placeholderMessage: 'Choose…',
            isRequired: true,
            requiredMessage: 'You need to select a tire size.',
          },
        },
        {
          key: 'brand',
          scope: 'public',
          includeForListingTypes: [
            'product-selling',
            'daily-booking',
            'nightly-booking',
            'hourly-booking',
          ],
          schemaType: 'enum',
          enumOptions: [
            { option: 'cube', label: 'Cube' },
            { option: 'diamant', label: 'Diamant' },
            { option: 'ghost', label: 'GHOST' },
            { option: 'giant', label: 'Giant' },
            { option: 'kalkhoff', label: 'Kalkhoff' },
            { option: 'kona', label: 'Kona' },
            { option: 'otler', label: 'Otler' },
            { option: 'vermont', label: 'Vermont' },
          ],
          filterConfig: {
            indexForSearch: true,
            label: 'Brand',
            group: 'secondary',
          },
          showConfig: {
            label: 'Brand',
            isDetail: true,
          },
          saveConfig: {
            label: 'Brand',
            placeholderMessage: 'Choose…',
            isRequired: true,
            requiredMessage: 'You need to select a brand.',
          },
        },
      ],
      listingTypes: [
        {
          listingType: 'rent-bicycles',
          label: 'Rent bicycles',
          transactionType: {
            process: 'default-booking',
            alias: 'default-booking/release-1',
            unitType: 'day',
          },
        },
        {
          listingType: 'product-selling',
          label: 'Sell products',
          transactionType: {
            process: 'default-purchase',
            alias: 'default-purchase/release-1',
            unitType: 'item',
          },
        },
      ],
    },
    search: {
      ...defaultConfig.search,
      mainSearch: {
        searchType: 'keywords',
      },
      defaultFilters: [defaultConfig.search.dateRangeFilter, defaultConfig.search.priceFilter],
    },
  };
};

export const getRouteConfiguration = () => {
  const layoutConfig = {
    searchPage: { variantType: 'map' },
    listingPage: { variantType: 'carousel' },
    listingImage: {
      aspectWidth: 400,
      aspectHeight: 400,
      variantPrefix: 'listing-card',
    },
  };
  return routeConfiguration(layoutConfig);
};

// Locale should not affect the tests. We ensure this by providing
// messages with the key as the value of each message.
const testMessages = mapValues(messages, (val, key) => key);

// Provide all the context for components that connect to the Redux
// store, i18n, router, etc.
export const TestProvider = ({ children }) => {
  const store = configureStore();
  return (
    <ConfigurationProvider value={getDefaultConfiguration()}>
      <RouteConfigurationProvider value={getRouteConfiguration()}>
        <IntlProvider locale="en" messages={testMessages} textComponent="span">
          <Provider store={store}>
            <HelmetProvider>
              <MemoryRouter>{children}</MemoryRouter>
            </HelmetProvider>
          </Provider>
        </IntlProvider>
      </RouteConfigurationProvider>
    </ConfigurationProvider>
  );
};

// With react-testing-library we don't have option to shallow render
// so we would need a custom render function to add required wrappers
// in order to test the whole Page component.
// Example for creating custom render function
// for using React Intl with RTL:
// https://testing-library.com/docs/example-react-intl#creating-a-custom-render-function
// function render(ui, { preloadedState, store = configureStore(), ...renderOptions } = {}) {
//   function Wrapper({ children }) {
//     return (<TestProvider>{children}</TestProvider>);
//   }
//   return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
// }

export const renderWithProviders = (ui, renderOptions = {}) => {
  const Wrapper = ({ children }) => {
    return <TestProvider>{children}</TestProvider>;
  };
  return reactTestingLibrary.render(ui, { wrapper: Wrapper, ...renderOptions });
};
export const testingLibrary = { ...reactTestingLibrary, userEvent };
