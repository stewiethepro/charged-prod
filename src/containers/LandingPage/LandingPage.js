import React from 'react';
import { bool, object } from 'prop-types';
import { compose } from 'redux';
import { connect } from 'react-redux';

import { camelize } from '../../util/string';
import { propTypes } from '../../util/types';

import PageBuilder from '../../containers/PageBuilder/PageBuilder';

import FallbackPage from './FallbackPage';
import { ASSET_NAME } from './LandingPage.duck';

import SectionFeaturesLandingHero from '../PageBuilder/SectionBuilder/SectionFeaturesLandingHero';
import BlockDefaultLandingHero from '../PageBuilder/BlockBuilder/BlockDefaultLandingHero';


export const LandingPageComponent = props => {
  const { pageAssetsData, inProgress, error, currentUser } = props;

  const sectionOverrides = {
    features: { component: SectionFeaturesLandingHero },
  };

  const blockOverrides = {
    defaultBlock: { component: BlockDefaultLandingHero },
  };

  // Boot Intercom

  if (typeof window !== "undefined") {
    if (currentUser && currentUser.attributes.emailVerified) {
      window.Intercom("boot", {
        api_base: "https://api-iam.intercom.io",
        app_id: "qv2ju58e",
        name: currentUser.attributes.profile.firstName + ' ' + currentUser.attributes.profile.lastName,
        user_id: currentUser.id.uuid, 
        email: currentUser.attributes.email,
        created_at: currentUser.attributes.createdAt
      });
    } else {
      window.Intercom("boot", {
        api_base: "https://api-iam.intercom.io",
        app_id: "qv2ju58e",
      });
    }
  }

  return (
    <PageBuilder
      pageAssetsData={pageAssetsData?.[camelize(ASSET_NAME)]?.data}
      inProgress={inProgress}
      options={{
        sectionComponents: sectionOverrides, 
        blockComponents: blockOverrides
      }}
      error={error}
      fallbackPage={<FallbackPage error={error}
      currentUser={currentUser} />}
    />
  );
};

LandingPageComponent.defaultProps = {
  currentUser: null,
};

LandingPageComponent.propTypes = {
  pageAssetsData: object,
  inProgress: bool,
  error: propTypes.error,
  currentUser: propTypes.currentUser,
};

const mapStateToProps = state => {
  const { pageAssetsData, inProgress, error } = state.hostedAssets || {};
  const { currentUser } = state.user;
  return { pageAssetsData, inProgress, error, currentUser };
};

// Note: it is important that the withRouter HOC is **outside** the
// connect HOC, otherwise React Router won't rerender any Route
// components since connect implements a shouldComponentUpdate
// lifecycle hook.
//
// See: https://github.com/ReactTraining/react-router/issues/4671
const LandingPage = compose(connect(mapStateToProps))(LandingPageComponent);

export default LandingPage;
