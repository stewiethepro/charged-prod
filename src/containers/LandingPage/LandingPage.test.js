import React from 'react';
import '@testing-library/jest-dom';

import { renderWithProviders as render } from '../../util/testHelpers';

import { LandingPageComponent } from './LandingPage';

describe('LandingPage', () => {
  // For now, we rely on snapshot-testing to check the content of fallback page.
  test('matches snapshot', () => {
    let error = new Error('Fake error');
    error.type = 'error';
    error.status = 404;
    error.type;

    // Testing that the fallback page looks as expected
    const tree = render(
      <LandingPageComponent pageAssetsData={{}} inProgress={false} error={error} />
    );
    expect(tree.asFragment().firstChild).toMatchSnapshot();
  });
});
