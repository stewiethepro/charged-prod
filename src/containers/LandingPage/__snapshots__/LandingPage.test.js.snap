// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LandingPage matches snapshot 1`] = `
<div
  class="root"
>
  <div
    class="content"
  >
    <div
      class="layout"
      style="grid-template-areas: 'topbar' 'main' 'footer'; display: grid;"
    >
      <header
        class="topbar"
        style="grid-area: topbar;"
      >
        <div
          class="root"
        >
          <div
            class="container"
          >
            <button
              class="menu"
              title="Topbar.menuIcon"
            >
              <svg
                class="rootMenuIcon menuIcon"
                height="12"
                viewBox="0 0 18 12"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill-rule="evenodd"
                >
                  <rect
                    height="2"
                    rx="1"
                    width="18"
                  />
                  <rect
                    height="2"
                    rx="1"
                    width="18"
                    y="5"
                  />
                  <rect
                    height="2"
                    rx="1"
                    width="18"
                    y="10"
                  />
                </g>
              </svg>
            </button>
            <a
              class="root NamedLink_active"
              href="/"
            >
              <img
                alt="Topbar.logoIcon"
                class="logo"
                src="biketribe-logo-mobile.png"
              />
            </a>
            <button
              class="searchMenu"
              title="Topbar.searchIcon"
            >
              <svg
                class="rootSearchIcon searchMenuIcon"
                height="18"
                viewBox="0 0 18 18"
                width="18"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g
                  fill="none"
                  fill-rule="evenodd"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  transform="matrix(-1 0 0 1 17 1)"
                >
                  <path
                    d="M11.733 11.733l3.727 3.727"
                  />
                  <circle
                    cx="6.4"
                    cy="6.4"
                    r="6.4"
                  />
                </g>
              </svg>
            </button>
          </div>
          <div
            class="desktop"
          >
            <nav
              class="root"
            >
              <a
                class="root logoLink NamedLink_active"
                href="/"
              >
                <img
                  alt="TopbarDesktop.logo"
                  class="logo"
                  src="biketribe-logo-desktop.png"
                />
              </a>
              <form
                action="/"
                class="searchLink"
                method="post"
              >
                <div
                  class="keywordSearchWrapper topbarSearchWithLeftPadding"
                >
                  <button
                    class="searchSubmit"
                  >
                    <div
                      class="desktopIcon"
                    >
                      <svg
                        class="iconSvg"
                        height="22"
                        viewBox="0 0 21 22"
                        width="21"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <g
                          class="iconSvgGroup"
                          fill="none"
                          fill-rule="evenodd"
                          stroke-linecap="round"
                          stroke-linejoin="round"
                          stroke-width="2"
                          transform="matrix(-1 0 0 1 20 1)"
                        >
                          <path
                            d="M13 14l5.241 5.241"
                          />
                          <circle
                            cx="7.5"
                            cy="7.5"
                            r="7.5"
                          />
                        </g>
                      </svg>
                    </div>
                  </button>
                  <input
                    autocomplete="off"
                    class="desktopInput"
                    id="keyword-search"
                    name="keywords"
                    placeholder="TopbarSearchForm.placeholder"
                    type="text"
                    value=""
                  />
                </div>
              </form>
              <a
                class="createListingLink"
                href="/l/new"
              >
                <span
                  class="createListing"
                >
                  <span>
                    TopbarDesktop.createListing
                  </span>
                </span>
              </a>
              <a
                class="signupLink"
                href="/signup"
              >
                <span
                  class="signup"
                >
                  <span>
                    TopbarDesktop.signup
                  </span>
                </span>
              </a>
              <a
                class="loginLink"
                href="/login"
              >
                <span
                  class="login"
                >
                  <span>
                    TopbarDesktop.login
                  </span>
                </span>
              </a>
            </nav>
          </div>
          <div
            class="genericError"
          >
            <div
              class="genericErrorContent"
            >
              <p
                class="genericErrorText"
              >
                <span>
                  Topbar.genericError
                </span>
              </p>
            </div>
          </div>
        </div>
      </header>
      <main
        class="main"
        style="grid-area: main;"
      >
        <section
          class="root"
          id="maintenance-mode"
        >
          <div
            class="content"
          >
            <h2>
              Maintenance mode
            </h2>
            <p>
              The marketplace is not fully operational at the moment.
              <br />
              Try refreshing the page and if that does not solve the issue, contact the marketplace admins.
            </p>
          </div>
        </section>
      </main>
      <div
        style="grid-area: footer;"
      >
        <div
          class="root"
        >
          <div
            class="topBorderWrapper"
          >
            <div
              class="content"
            >
              <div
                class="socialLinksMobile"
              />
              <div
                class="links"
              >
                <div
                  class="organization"
                  id="organization"
                >
                  <a
                    class="logoLink NamedLink_active"
                    href="/"
                  >
                    <img
                      alt="MarketplaceX"
                      class="logo"
                      src="biketribe-logo-desktop.png"
                    />
                  </a>
                  <div
                    class="organizationInfo"
                  >
                    <p
                      class="organizationDescription"
                    >
                      <span>
                        Footer.organizationDescription
                      </span>
                    </p>
                    <p
                      class="organizationCopyright"
                    >
                      <a
                        class="copyrightLink NamedLink_active"
                        href="/"
                      >
                        <span>
                          Footer.copyright
                        </span>
                      </a>
                    </p>
                  </div>
                </div>
                <div
                  class="infoLinks"
                >
                  <ul
                    class="list"
                  >
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/p/about"
                      >
                        <span>
                          Footer.toAboutPage
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s"
                      >
                        <span>
                          Footer.toSearchPage
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/l/new"
                      >
                        <span>
                          Footer.toNewListingPage
                        </span>
                      </a>
                    </li>
                  </ul>
                </div>
                <div
                  class="searches"
                >
                  <ul
                    class="list"
                  >
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Helsinki%2C%20Finland&bounds=60.2978389%2C25.254484899999966%2C59.9224887%2C24.782875800000056&origin=60.16985569999999%2C24.93837910000002"
                      >
                        <span>
                          Footer.link1
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Turku%2C%20Finland&bounds=60.5304531%2C22.38197820000005%2C60.33361189999999%2C22.066440100000023&origin=60.4518126%2C22.26663020000001"
                      >
                        <span>
                          Footer.link2
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Tampere%2C%20Finland&bounds=61.83657359999999%2C24.118384099999957%2C61.42728169999999%2C23.542201299999988&origin=61.4977524%2C23.760953500000028"
                      >
                        <span>
                          Footer.link3
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Oulu%2C%20Finland&bounds=65.5643426%2C26.770696000000044%2C64.8443082%2C24.114941999999928&origin=65.0120888%2C25.465077199999996"
                      >
                        <span>
                          Footer.link4
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Rukatunturi%2C%20Kuusamo%2C%20Finland&bounds=66.1699707%2C29.167735499999935%2C66.1609529%2C29.13572069999998&origin=66.1654622%2C29.151728100000014"
                      >
                        <span>
                          Footer.link5
                        </span>
                      </a>
                    </li>
                  </ul>
                </div>
                <div
                  class="searchesExtra"
                >
                  <ul
                    class="list"
                  >
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Kuopio%2C%20Finland&bounds=63.40340329999999%2C28.457629300000008%2C62.592254%2C26.890028099999995&origin=62.89796999999999%2C27.67817249999996"
                      >
                        <span>
                          Footer.link6
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Mikkeli%2C%20Finland&bounds=62.1990357%2C27.900274999999965%2C61.20955799999999%2C26.82762809999997&origin=61.68872709999999%2C27.27214570000001"
                      >
                        <span>
                          Footer.link7
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Espoo%2C%20Finland&bounds=60.3636105%2C24.8505715%2C60.0499087%2C24.499656500000015&origin=60.2054911%2C24.655899999999974"
                      >
                        <span>
                          Footer.link8
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Vantaa%2C%20Finland&bounds=60.40113659999999%2C25.19306240000003%2C60.23714200000001%2C24.746512199999984&origin=60.29335239999999%2C25.037768599999936"
                      >
                        <span>
                          Footer.link9
                        </span>
                      </a>
                    </li>
                    <li
                      class="listItem"
                    >
                      <a
                        class="link"
                        href="/s?address=Ahvenanmaan%20saaristo%2C%20Åland%20Islands&bounds=60.87664400000001%2C21.34532850000005%2C59.45425059999999%2C19.94061499999998&origin=59.8994405%2C20.593996100000027"
                      >
                        <span>
                          Footer.link10
                        </span>
                      </a>
                    </li>
                  </ul>
                </div>
                <div
                  class="extraLinks"
                >
                  <div
                    class="legalMatters"
                  >
                    <ul
                      class="tosAndPrivacy"
                    >
                      <li>
                        <a
                          class="legalLink"
                          href="/terms-of-service"
                        >
                          <span>
                            Footer.termsOfUse
                          </span>
                        </a>
                      </li>
                      <li>
                        <a
                          class="legalLink"
                          href="/privacy-policy"
                        >
                          <span>
                            Footer.privacyPolicy
                          </span>
                        </a>
                      </li>
                    </ul>
                  </div>
                </div>
              </div>
              <div
                class="copyrightAndTermsMobile"
              >
                <a
                  class="organizationCopyrightMobile NamedLink_active"
                  href="/"
                >
                  <span>
                    Footer.copyright
                  </span>
                </a>
                <div
                  class="tosAndPrivacyMobile"
                >
                  <a
                    class="privacy"
                    href="/privacy-policy"
                  >
                    <span>
                      Footer.privacy
                    </span>
                  </a>
                  <a
                    class="terms"
                    href="/terms-of-service"
                  >
                    <span>
                      Footer.terms
                    </span>
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
